### PR TITLE
Way for API to announce any extensions they use in addition to Hydra Core

### PR DESCRIPTION
**Pull request by alien-mcl**
Introduction of the `hydra:extension` predicate that should resolve issue #221

## Summary

This PR should resolve issue #221 by introducing a new term by which server could announce extensions in use. Additionally, client can express it's preferences regarding those extensions.

## More details

While in the issue there was a suggestion of using a `hydra:profile` term, I decided to use `hydra:extension`. The word `profile` introduced was meant to reflect how descriptive logic of OWL changes with different OWL profiles, thus it didn't feel like a correct word to use. Finally, _extension_ feels more generic and does not imply any specific behavior other than _something additional to the base Hydra description_ that client may or may not understand.

### CODE REVIEW

File: spec/latest/core/index.html [hunk 1]
```diff
@@ -1287,6 +1287,37 @@ <h3>Client initiated pagination</h3>
...
       instead of a number.</p>
   </section>
+
+  <section>
+    <h3>Extensions</h3>
+    <p>While Hydra Core Vocabulary allows to address many usage scenarios,
+      not every aspect of API behavior can be covered with it. This
+      applies especially to querying, resource projection or data structure
+      description. This is due to fact that Hydra was meant to be as light
+      as possible forcing to drop some of the features out of it; scope.</p>
```
Comment (tpluscode) [inline spec/latest/core/index.html:1297]:
```suggestion
      as possible forcing to drop some of the features out of its scope.</p>
```
?

File: spec/latest/core/index.html [hunk 2]
```diff
@@ -1287,6 +1287,37 @@ <h3>Client initiated pagination</h3>
...
+      additional vocabularies.</p>
+
+    <p>It is up to the used vocabulary how these additional details should be
+      interpreted and in case client does not recognize these extensions,
+      additional details should be ignored and base Hydra interpretation
+      should be in force.</p>
+
+    <p>Server SHOULD NOT use extensions to add statements that are in
+      contradiction to base Hydra interpretation so the client is not confused.
+      Server also SHOULD keep multiple extensions describing adequate
```
Comment (tpluscode) [inline spec/latest/core/index.html:1312]:
```suggestion
      Server SHOULD also keep multiple extensions describing adequate
```

File: spec/latest/core/index.html [hunk 3]
```diff
@@ -1287,6 +1287,37 @@ <h3>Client initiated pagination</h3>
...
+      knowledge in line regarding their description (i.e. data structure
+      descriptions in various vocabularies should not cause differences).</p>
+
+    <p>Client can express it's preferences through the <i>PREFER</i> header
```
Comment (tpluscode) [inline spec/latest/core/index.html:1316]:
```suggestion
    <p>Client can express its preferences through the <i>PREFER</i> header
```

**Review by tpluscode — CHANGES_REQUESTED** [review]

**Comment by alien-mcl** [comment]
I've updated PR with more detailed PREFER header construction.

File: spec/latest/core/index.html [hunk 4]
```diff
@@ -1287,6 +1287,51 @@ <h3>Client initiated pagination</h3>
       to provide a custom page identifier (i.e. a GUID or a letter)
       instead of a number.</p>
   </section>
+
+  <section>
+    <h3>Extensions</h3>
+    <p>While Hydra Core Vocabulary allows to address many usage scenarios,
```
Comment (tpluscode) [inline spec/latest/core/index.html:1293]:
```suggestion
    <p>While Hydra Core Vocabulary allows addressing many usage scenarios,
```

File: spec/latest/core/index.html [hunk 5]
```diff
@@ -1287,6 +1287,51 @@ <h3>Client initiated pagination</h3>
...
+      not every aspect of API behavior can be covered with it. This
```
Comment (tpluscode) [inline spec/latest/core/index.html:1294]:
```suggestion
      not every aspect of API behavior can be covered. This
```

File: spec/latest/core/index.html [hunk 6]
```diff
@@ -1287,6 +1287,51 @@ <h3>Client initiated pagination</h3>
...
+      applies especially to querying, resource projection or data structure
+      description. This is due to fact that Hydra was meant to be as light
+      as possible forcing to drop some of the features out of its scope.</p>
+
+    <p>That is why there is a possibility of hinting a client on what kind
+      of extensions may be found or used in the received payloads.
+      After discovering a <i>extension</i> predicate in the API documentation,
```
Comment (tpluscode) [inline spec/latest/core/index.html:1301]:
```suggestion
      After discovering an <i>extension</i> predicate in the API documentation,
```

File: spec/latest/core/index.html [hunk 7]
```diff
@@ -1287,6 +1287,51 @@ <h3>Client initiated pagination</h3>
...
+      client can assume additional details are available described with
+      additional vocabularies.</p>
```
Comment (tpluscode) [inline spec/latest/core/index.html:1303]:
```suggestion
      client can assume additional details are available, described with
      complementary vocabularies.</p>
```

**Review by tpluscode — APPROVED** [review]
Just some editorial changes. Otherwise LGTM

File: spec/latest/core/index.html [hunk 8]
```diff
@@ -1287,6 +1287,51 @@ <h3>Client initiated pagination</h3>
       to provide a custom page identifier (i.e. a GUID or a letter)
       instead of a number.</p>
   </section>
+
+  <section>
+    <h3>Extensions</h3>
+    <p>While Hydra Core Vocabulary allows to address many usage scenarios,
+      not every aspect of API behavior can be covered with it. This
+      applies especially to querying, resource projection or data structure
+      description. This is due to fact that Hydra was meant to be as light
```
Comment (dunglas) [inline spec/latest/core/index.html:1296]:
```suggestion
      description. This is due to fact that Hydra is meant to be as light
```

File: spec/latest/core/index.html [hunk 9]
```diff
@@ -1287,6 +1287,51 @@ <h3>Client initiated pagination</h3>
...
+      as possible forcing to drop some of the features out of its scope.</p>
+
+    <p>That is why there is a possibility of hinting a client on what kind
+      of extensions may be found or used in the received payloads.
+      After discovering a <i>extension</i> predicate in the API documentation,
+      client can assume additional details are available described with
+      additional vocabularies.</p>
+
+    <p>It is up to the used vocabulary how these additional details should be
```
Comment (dunglas) [inline spec/latest/core/index.html:1305]:
```suggestion
    <p>It is up to the used vocabulary to define how these additional details should be
```

File: spec/latest/core/index.html [hunk 10]
```diff
@@ -1287,6 +1287,51 @@ <h3>Client initiated pagination</h3>
...
+      interpreted and in case client does not recognize these extensions,
```
Comment (dunglas) [inline spec/latest/core/index.html:1306]:
```suggestion
      interpreted. In case client does not recognize these extensions,
```

File: spec/latest/core/index.html [hunk 11]
```diff
@@ -1287,6 +1287,51 @@ <h3>Client initiated pagination</h3>
...
+      additional details should be ignored and base Hydra interpretation
+      should be in force.</p>
+
+    <p>Server SHOULD NOT use extensions to add statements that are in
+      contradiction to base Hydra interpretation so the client is not confused.
+      Server SHOULD also keep multiple extensions describing adequate
+      knowledge in line regarding their description (i.e. data structure
+      descriptions in various vocabularies should not cause differences).</p>
+
+    <p>Client can express its preferences through the <i>PREFER</i> header
```
Comment (dunglas) [inline spec/latest/core/index.html:1316]:
```suggestion
    <p>Client can express its preferences through the <i>Prefer</i> HTTP header
```

File: spec/latest/core/index.html [hunk 12]
```diff
@@ -1287,6 +1287,51 @@ <h3>Client initiated pagination</h3>
...
+      by pointing the preferred extensions via IRIs as on the example below.
+      The expectation of <i>hydra.extension</i> should be followed by a
```
Comment (dunglas) [inline spec/latest/core/index.html:1318]: You could maybe just refer to the ABNF definition providing by the RFC instead of trying to describe the syntax in plain English? The example should be enough for most readers:

```suggestion
    The client SHOULD use the `Prefer` HTTP header [[!RFC7240]] with the `hydra.extension` preference an an <i>iri</i> attribute having the IRI of the extension as value to hint the server about the extensions it supports.
```

File: spec/latest/core/index.html [hunk 13]
```diff
@@ -1287,6 +1287,51 @@ <h3>Client initiated pagination</h3>
...
+
+    <pre class="example nohighlight" data-transform="updateExample"
+         title="Hydra client extension preference.">
+      <!--
+      GET http://api.example.com/api/people
+      Prefer: hydra.extension; iri="http://schema.org/"
+      -->
+    </pre>
+
+    <p>Server SHOULD implement <i>PREFER</i> header handling according to
```
Comment (dunglas) [inline spec/latest/core/index.html:1332]:
```suggestion
    <p>Server MUST implement <i>Prefer</i> header handling according to
```

Implementing properly the RFC should be mandatory, isn't it?

**Review by dunglas — APPROVED** [review]
I made minor comments, but overall it looks good to me! Thanks for working on this.

**Review by tpluscode — APPROVED** [review]

File: spec/latest/core/index.html [hunk 14]
```diff
@@ -1287,6 +1287,51 @@ <h3>Client initiated pagination</h3>
...
+      applies especially to querying, resource projection or data structure
+      description. This is due to fact that Hydra is meant to be as light
+      as possible forcing to drop some of the features out of its scope.</p>
+
+    <p>That is why there is a possibility of hinting a client on what kind
+      of extensions may be found or used in the received payloads.
+      After discovering an <i>extension</i> predicate in the API documentation,
+      client can assume additional details are available described with
+      complementary vocabularies.</p>
+
```
Comment (alien-mcl) [inline spec/latest/core/index.html:1304]: Add an example hinting usage of some extension, i.e. vocabulary for data querying.

**Review by tpluscode — APPROVED** [review]
